js-pkg/packages/driftdb-react/src/index.tsx

```diff
@@ -13,59 +13,104 @@ import React, { useCallback, useEffect, useRef, useState } from 'react'
 
 const ROOM_ID_KEY = '_driftdb_room'
 
-export const DatabaseContext = React.createContext<DbConnection | null>(null)
-
-export function useDatabase(): DbConnection {
-  const db = React.useContext(DatabaseContext)
-  if (db === null) {
-    throw new Error('useDatabase must be used within a DriftDBProvider')
+/**
+ * A React component that provides a `DbConnection` to all child components.
+ * 
+ * @param props The props for the component.
+ */
+export function DriftDBProvider(props: {
+  /** Elements under the prodider in the tree. */
+  children: React.ReactNode
+  /** The URL of the DriftDB API. */
+  api: string
+  /** The room ID to connect to. If not provided, attempts to extract the room ID
+   *  from the URL and creates a new room if one is not present. */
+  room?: string
+  /** Whether to use binary messages (enables raw typed arrays in messages). */
+  useBinary?: boolean
+}): React.ReactElement {
+  const dbRef = useRef<DbConnection | null>(null)
+  if (dbRef.current === null) {
+    dbRef.current = new DbConnection()
   }
-  return db
-}
 
-export function RoomQRCode() {
-  const db = useDatabase()
-  const [pageUrl, setPageUrl] = React.useState<string | null>(null)
+  React.useEffect(() => {
+    let api = new Api(props.api)
 
-  useEffect(() => {
-    const callback = () => {
-      if (typeof window === 'undefined') {
-        return
-      }
+    let roomId
+    if (props.room) {
+      roomId = props.room
+    } else {
+      const searchParams = new URLSearchParams(window.location.search)
+      roomId = searchParams.get(ROOM_ID_KEY)
+    }
 
-      const url = new URL(window.location.href)
-      const checkRoom = url.searchParams.get(ROOM_ID_KEY)
+    let promise
+    if (roomId) {
+      promise = api.getRoom(roomId)
+    } else {
+      promise = api.newRoom()
+    }
 
-      if (!checkRoom) {
-        return
+    promise.then((result: RoomResult) => {
+      if (!props.room) {
+        let url = new URL(window.location.href)
+        url.searchParams.set(ROOM_ID_KEY, result.room)
+        window.history.replaceState({}, '', url.toString())
       }
```

Comment:
Not really a blocker, but something we should address eventually: this promise callback can fire after the component has been unmounted and `db.disconnect()` has been called. Typically, this is avoided by using a "cancellable promise". As far as I know, you have to implement that yourself right now? You can use an approach like [this](https://dev.to/praveenkumarrr/cancellable-promises-in-react-and-why-is-it-required-5ghf).

Reply:
good catch, created https://github.com/drifting-in-space/driftdb/issues/73